reactioncommerce/reaction
# Remove account from anon carts, don't use session for carts

Resolves #4482 
Resolves #4464 
Impact: **breaking**  
Type: **feature**

## Issue
There were a few bugs found with management and merging of anonymous carts after the changes in #4412 were merged.

## Solution
Unfortunately, these bugs highlighted the need to more fully refactor cart management, which overlapped with the idea of Reaction "sessions", and so sessions are also refactored. Carts will now be handled the way we are handling them in the new NextJS starter kit app.

## Changes
- Refactor `getCart` to be different on client and server because the needs are different (same version change)
- [BREAKING] A cart is not created until items are added. Previously a cart was created for all users, including anonymous users, immediately if one was not found. This is not a breaking change for the core app, but any custom plugins may have code that will need to be updated to handle the possibility of there not being a cart.
- [BREAKING] Update the signature of most cart methods to take an optional `cartToken` string param. Update all places that call these methods to pass in the token for anonymous carts.
- Client code now tracks anonymous carts (up to one per shop) in local storage (fallback to cookie). The storage key is "Reaction.anonymousCarts", which is an array of objects with `_id`, `token`, and `shopId` properties. All interaction with this data should be done by importing the util functions from `"/imports/plugins/core/cart/client/util/anonymousCarts"`
- [BREAKING] The "Reaction.sessionId" stored ID is now used only for auto-login of anonymous users. It is not used by any of the cart code. Also, the "Sessions" collection is no longer written to or published to clients. It will not be dropped automatically, but you can drop it if you no longer need it.
- [BREAKING] `Accounts.loginWithAnonymous` is no longer available to client code. This was only used in one place, and similar logic has replaced it in that spot.
- The client autorun that calls `mergeCart` is now smarter about calling it only when necessary. (same version fix)
- While I was in the file, I fixed #4464 by using a different applicable message that already exists.
- `addressBookRemove` method no longer clears the removed address from any carts. We've decided that addresses in cart should stay separated from address book once copied into the cart.
- [BREAKING] `workflow/pushCartWorkflow` and `workflow/revertCartWorkflow` methods now require that you pass in the `cartId` rather than guessing which cart you intend.
- [BREAKING] In general, be aware that `cart.accountId` may now be null. Previously, it would be set even for anonymous carts, to the account for the user with "anonymous" role. For now, `order.accountId` is still set after an anonymous order is placed.

## Testing
This requires thorough testing of the following:
- Adding your first item to cart when not logged in
- Adding another item to cart when not logged in
- Adding your first item to cart when logged in
- Adding another item to cart when logged in
- Guest checkout through to thank you screen with flat rate shipping and example payment
- Guest checkout through to thank you screen with flat rate shipping and Stripe payment
- Logged in checkout through to thank you screen with flat rate shipping and example payment
- Logged in checkout through to thank you screen with flat rate shipping and Stripe payment
- Logging in during checkout
- Having items in anonymous cart and an account with no cart, log in and items should be transferred to account cart. Then log out and cart should be empty.
- Having items in anonymous cart and an account with a cart with some items, log in and items should be transferred to account cart, merged with existing items (including merging quantity if the same item is in both carts).
- Checking out twice in a row as anonymous (add items, checkout, add more items, checkout again)
- Checking out twice in a row as logged in
